drivers/sensor/max17055/max17055.c

```diff
@@ -0,0 +1,244 @@
+		tmp = capacity_to_ma(config->rsense_mohms, priv->design_cap);
+		set_millis(val, tmp);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIGN_VOLTAGE:
+		set_millis(val, config->design_voltage);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIRED_VOLTAGE:
+		set_millis(val, config->desired_voltage);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT:
+		val->val1 = config->desired_charging_current;
+		val->val2 = 0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int max17055_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct max17055_data *priv = dev->driver_data;
+	int rc = 0;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	rc = max17055_reg_read(priv, VCELL, &priv->voltage);
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, AVG_CURRENT, &priv->avg_current);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, REP_SOC, &priv->state_of_charge);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, INT_TEMP, &priv->internal_temp);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, REP_CAP, &priv->remaining_cap);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, FULL_CAP_REP, &priv->full_cap);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, TTE, &priv->time_to_empty);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, TTF, &priv->time_to_full);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, CYCLES, &priv->cycle_count);
+	}
+	if (rc == 0) {
+		rc = max17055_reg_read(priv, DESIGN_CAP, &priv->design_cap);
+	}
+	if (rc != 0) {
+		LOG_ERR("Failed to read channel %d", chan);
+		return rc;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief initialise the fuel gauge
+ *
+ * @return 0 for success
+ * @return -EINVAL if the I2C controller could not be found
+ */
+static int max17055_gauge_init(struct device *dev)
+{
+	struct max17055_data *priv = dev->driver_data;
+	const struct max17055_config *const config = dev->config_info;
+
+	priv->i2c = device_get_binding(config->bus_name);
+	if (!priv->i2c) {
+		LOG_ERR("Could not get pointer to %s device", config->bus_name);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static const struct sensor_driver_api max17055_battery_driver_api = {
+	.sample_fetch = max17055_sample_fetch,
+	.channel_get = max17055_channel_get,
+};
+
+#define MAX17055_INIT(index)						\
+	static struct max17055_data max17055_driver_##index;		\
+									\
+	static const struct max17055_config max17055_config_##index = { \
+		.bus_name = DT_INST_BUS_LABEL(index),			\
+		.rsense_mohms = DT_INST_PROP(index, rsense_mohms),	\
+		.design_voltage = DT_INST_PROP(index, design_voltage),	\
+		.desired_voltage = DT_INST_PROP(index, desired_voltage), \
+		.desired_charging_current =				\
+			DT_INST_PROP(index, desired_charging_current),	\
+	};								\
+									\
+	DEVICE_AND_API_INIT(max17055_##index, DT_INST_LABEL(index),	\
+			    &max17055_gauge_init, &max17055_driver_##index, \
+			    &max17055_config_##index, POST_KERNEL,	\
+			    CONFIG_SENSOR_INIT_PRIORITY,		\
+			    &max17055_battery_driver_api);
```

Comment:
Use the macro with a semicolon at the callsite, instead of adding to the end of the macro

Reply:
I can do that only if the macro is instantiated once. Of course it would be odd to have two chargers. But I see that other drivers in Zephyr avoid having a semicolon in the macro

Very rough check:

git grep -A7 DEVICE_AND_API_INIT |grep ");" |wc

git grep -A7 DEVICE_AND_API_INIT |grep ")$" |wc